[Feat] Jit warp perspective

kuynzereb opened the pull request:
### Description
I made some refactoring of the homography warping functions, so the `warp_perspective()` is now can be torchscripted. The main difference is that now warping implemented in `homography_warp()` and  not in `HomographyWarper` class, so `transform_warp_impl()` can call directly to `homography_warp()` without creation of `HomographyWarper` instance.

As a side effect `warp_grid()` is now a global function (previously it was `HomographyWarper` method) and `homography_warp()` has an additional argument `normalized_coordinates`. It is quite strange that it was not there before, because `HomographyWarper` accept this argument.

### Status
Ready

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [x] Docstrings/Documentation updated


## PR Checklist
### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
The only failed test is:
```
FAILED test/geometry/transform/test_pyramid.py::TestPyrDown::test_symmetry_preserving[cuda] - AssertionError: assert False
```
I guess, it is not related to the changes from this PR
- [ ] Unittests: did you add tests for your new functionality ?
- [x] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

edgarriba commented:
@kuynzereb this is great! just one missing thing, no test using the jitted version that verifies this works. Besides, notice that there's an on going effort to make a generic API for warping to allow any kinds of transforms (rigid/non-rigid/2D/3D) using grid sample as a backend. Check it out, and feel free to take this task or give some ideas: https://github.com/kornia/kornia/issues/563 /cc @lferraz @mibaumgartner

kuynzereb commented:
I have added the tests to check that jit versions of `warp_homography()` and `warp_perspective()` give the same results as pytorch versions. Also I have fixed the homography warper test, where random homographies always were trivial because random noise was initialized with zeros.

lferraz commented on kornia/geometry/warp/homography_warper.py:
```diff
@@ -36,6 +112,7 @@ class HomographyWarper(nn.Module):
         align_corners(bool): interpolation flag. Default: False. See
         https://pytorch.org/docs/stable/nn.functional.html#torch.nn.functional.interpolate for detail
     """
+    _warped_grid: Optional[torch.Tensor]
```
why do you define the variable as a class variable. I have the feeling that it is dangerous. why not an instance variable?

kuynzereb replied:
It is not a class variable, it is just a typing annotation: https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-annotations

It is required if we want the instance of this class to be torchscriptable. See “Module Attributes” section of https://pytorch.org/docs/stable/jit_language_reference.html#language-reference for details.

It was not the purpose of this PR, but I have made it so in the future it is easier to make the instances of this class torchscriptable.

lferraz replied:
ok, I never did the typing there but if it is a requirement of jit its fine.

lferraz commented on kornia/geometry/warp/homography_warper.py:
```diff
@@ -71,35 +148,9 @@ def precompute_warp_grid(self, src_homo_dst: torch.Tensor) -> None:
-        return flow.view(batch_size, self.height, self.width, 2)  # NxHxWx2
-
-    def forward(  # type: ignore
+    def forward(
```
regarding the  `# type: ignore`, did they correct the issue of mypy with the forward method? I will be very happy :)

kuynzereb replied:
I don't really know what do you mean :)
I removed it because jit gave errors because of it

lferraz replied:
I do not remember why, but mypy hates the forward function. If it passes the test it is fine.

edgarriba replied:
great! good to see that they fixed this :)

edgarriba commented on kornia/geometry/warp/homography_warper.py:
```diff
@@ -6,16 +6,92 @@
 ]
 
 
+def warp_grid(grid: torch.Tensor, src_homo_dst: torch.Tensor) -> torch.Tensor:
```
probably it's out of the scope o this PR but at as comment for future refactoring whoever takes it, this function doesn't compute the warping for a grid it just create it as `torch.nn.functional.affine_grid` already does. We should explore to use `affine_grid` for the homography case since e.g. in 2d affine transforms we saw that the speed for this simple function using torch cpp (affine_grid) improves by far the speed performnance.

edgarriba commented on kornia/geometry/warp/homography_warper.py:
```diff
@@ -129,67 +180,28 @@ def forward(  # type: ignore
+                                 recall precompute_warp_grid() with the correct device \
+                                 for the homograhy or change the patch device.".format(
+                                patch_src.device, _warped_grid.device))
+            warped_patch = F.grid_sample(
```
probably to encapsulate somewhere `grid_sample` we could replace this call to `remap`

edgarriba commented:
@kuynzereb thanks for the contribution, and sorry for late merge.

kuynzereb commented:
Great, no problem :)
